text/0000-clustering.md

```diff
@@ -0,0 +1,218 @@
+[guide-level-explanation]: #guide-level-explanation
+
+Tremor can be operated on a clustered mode, which allows operators to use a
+group of tremor nodes collaboratively for their event processing needs, giving
+the illusion of a single running tremor instance to the outside world.
+
+A tremor cluster is composed of two sets of nodes: `coordinators` and `workers`.
+
+Coordinator nodes are the ones coordinating the addition and removal of nodes to
+the cluster. They also intercept all changes to a running cluster (basically new
+workload submissions, or changes to the tremor [repository and registry
+state](https://docs.tremor.rs/operations/configuration/#introduction)) and
+ensure that all nodes eventually apply the changes -- anytime the [tremor
+api](https://docs.tremor.rs/api/) used to submit changes from a node, the
+request is routed first to the coordinator node (if the node accepting the
+request is not already a coordinator). Thus, the group of coordinator nodes can
+be thought of as the brain of the cluster. For the initial clustering
+implementation, the allowed number of dedicated coordinator nodes is 3.
+
+If coordinators form the brain, worker nodes are the brawn of the cluster,
+hosting the actual running instances of the deployed sources, pipelines and
+sinks. Worker nodes accept changes from the coordinator nodes for this and
+communicate directly to the cluster-external world as their workload demands
+(i.e. as part of the sources/sinks running on the node). There is no set limit
+on the number of worker nodes in the cluster, though it will be bound in
+practicality by the network constraints of cluster-wide communication.
+
+An example of starting a tremor server process in these two modes:
+
+```sh
+# assuming this command is run on host1 and the 3 coordinator nodes are on
+# host1, host2, host3 (host1 can be left out here if desired)
+tremor server run --coordinator --peers "host1:8139,host2:8139,host3:8139"
+
+# assuming this command is run on host4 (all 3 coordinator nodes have to be specified)
+tremor server run --worker --coordinators "host1:8139,host2:8139,host3:8139"
+```
+
+The static host list specified above is the current means of cluster discovery
+(i.e. how a node finds the tremor cluster to join). The use of `--coordinator`
+and `--worker` flag mandates the use of `--peers` and `--coordinators`
+respectively -- the server will exit with an error during initialization if one
+is used without the other. The default port for cluster communication is 8139
+(if port is left-out as part of host details above, that’s the assumed port),
+and can be overridden by a `--cluster-port` flag for both coordinators and
+workers. Example:
+
+```sh
+# cluster-port flag is relevant only when coordinator or worker flag is specified
+tremor server run --coordinator --peers "host1:8139,host2:8139,host3:8139" --cluster-port 4242
+```
+
+When started without these flags, the server process simulates the behavior of
```

Comment:
Lets revisit this and see if overloading server run makes sense verses adding `tremor cluster run ...` later. Maintaining standalone mode of operation is well captured and important for some use cases.